orderbook/src/database/orders.rs

```diff
@@ -261,14 +286,85 @@ impl OrdersQueryRow {
 mod tests {
 
     use super::*;
-    use chrono::NaiveDateTime;
+    use chrono::{Duration, NaiveDateTime};
     use futures::StreamExt;
     use num_bigint::BigUint;
     use primitive_types::U256;
     use shared::event_handling::EventIndex;
     use sqlx::Executor;
     use std::collections::HashSet;
 
+    #[test]
+    fn order_status() {
```

Comment:
This test could probably include an open, closed and expired partially fillable order. Starting to make me think that we might want the status to be a tuple

(ValidityStatus, FillStatus)

Where 

Validity is any of (open, expired) and FillStatus is any of (not filled, partially filled or fully filled)

From this you can also derive stuff like closed.

Reply:
I agree that fill status and validity are kind of orthogonal (not even sure we need to indicate fill status as an enum, since we already expose the filled amount which should make it trivial to identify if an order is not, partially or fully filled).

I think it might still make sense to have the ValidityStatus also include something like the Fulfilled suggested here, because that is another way of how an order can become invalid (smart contract would complain with AlreadyFilled).

Reply:
From my view, the order has 4 states, where 1 is transient (`Open`) and 3 are final (`Fulfilled`, `Cancelled`, `Expired`).
Or you can call the last 3 `Closed` states, for example:

```rust

enum ClosedStates {
    Fulfilled,
    Cancelled,
    Expired,
}

enum OrderState {
     Open,
     Closed(ClosedStates),
}
```

To keep it simple though, I still prefer the flat enum. Maybe would make more sense if we had more `Open` states, like `Pending` or something, but not as is.

Orthogonal to that, there's the fill state, which as Ben mentioned are `Not filled`, `Partially filled` and `Fully filled`.
`Not filled` is implied, as this is the initial state for all orders.
`Fully filled` only applies to the `Closed(Fulfilled)` state.
`Partially filled` can apply to `Open`, `Cancelled` and `Expired`.

Summarizing, I think we can achieve what Ben suggests by keeping the states as I propose, and adding a boolean flag to the Order model called: `PartiallyFilled`

What do you think @bh2smith ?

Reply:
Side note, for the test specifically, I've added a few more covering orders with partial fills https://github.com/gnosis/gp-v2-services/pull/715/commits/ff7c6336586882d495ad1307749098217bffbb52